### PR TITLE
feat(gapic-generator-ruby): introduce auto_populated_fields (uuid4) to generator

### PR DESCRIPTION
**Pull request by aandreassa**

### CODE REVIEW

File: gapic-generator/templates/default/service/client/method/def/_request_streaming.erb [hunk 1]
```diff
@@ -4,6 +4,17 @@ unless request.is_a? ::Enumerable
...
+if request.is_a? Hash
+  request[:<%= field %>] = SecureRandom.uuid if request.key?(:<%= field %>) && request[:<%= field %>].empty?
+else
+  request.<%= field %> = SecureRandom.uuid if request.respond_to?(:<%= field %>) && request.<%= field %>.empty?
+end
+<% end %>
 request = request.lazy.map do |req|
   ::Gapic::Protobuf.coerce req, to: <%= method.request_type %>
-end
+end
```
Comment (dazuma) [inline gapic-generator/templates/default/service/client/method/def/_request_streaming.erb:20]: Don't remove the trailing newline.

**Review by dazuma — CHANGES_REQUESTED** [review]

File: gapic-generator/lib/gapic/model/api_metadata.rb [hunk 2]
```diff
@@ -138,6 +152,21 @@ def remove_html_tags str
           .gsub(%r{</?[[:alpha:]]+(?:\s+[[:alpha:]]+\s*=\s*(?:"[^"]+"|'[^']+'|[^\s">]+))*\s*/?>}, "")
           .gsub(%r{\[([^\]]+)\]\([^)]+\)}, "\\1")
       end
+
+      def auto_populated_methods! service_config
```
[the remark this answers is not in the document]
Comment (aandreassa) [reply gapic-generator/lib/gapic/model/api_metadata.rb]: Done, removed `method_settings` + added private to `standardize_auto_populated_fields!`

File: gapic-generator/lib/gapic/model/api_metadata.rb [hunk 3]
```diff
@@ -70,7 +70,21 @@ class ApiMetadata
       #
       attr_reader :doc_tag_prefix
 
+      ##
+      # @return [Hash{String => Hash{String => Array<String>}}] A mapping from method selector
+      #     to individual method settings (e.g. auto_populated_fields).
+      #
+      attr_reader :method_settings
```
[the remark this answers is not in the document]
Comment (aandreassa) [reply gapic-generator/lib/gapic/model/api_metadata.rb]: Removed

File: gapic-generator/lib/gapic/model/api_metadata.rb [hunk 4]
```diff
@@ -70,7 +70,21 @@ class ApiMetadata
...
+
+      ##
+      # @return [Hash{String => Array<String>}] A mapping from method name to a list of auto
+      #     populated fields.
+      #
+      attr_reader :auto_populated_methods
```
[the remark this answers is not in the document]
Comment (aandreassa) [reply gapic-generator/lib/gapic/model/api_metadata.rb]: Used `auto_populated_fields_by_method_name`

File: gapic-generator/lib/gapic/model/api_metadata.rb [hunk 5]
```diff
@@ -138,6 +152,21 @@ def remove_html_tags str
...
+        @method_settings.each do |setting|
+          selector = setting[SELECTOR]
+          methods = service_config.apis.filter_map do |api|
+            # Removes API prefix and trailing period.
+            selector[api.name.length + 1..].downcase if selector.start_with? api.name
+          end
+          if !methods.nil? && setting.key?(AUTO_POPULATED_FIELDS)
+            @auto_populated_methods[methods.first] = setting[AUTO_POPULATED_FIELDS]
+          end
+        end
```
[the remark this answers is not in the document]
Comment (aandreassa) [reply gapic-generator/lib/gapic/model/api_metadata.rb:169]: I kept it consistent with same format above (return unless ... & self)

File: gapic-generator/templates/default/service/client/method/def/_request_streaming.erb (hunk 1 above)
Comment (aandreassa) [reply gapic-generator/templates/default/service/client/method/def/_request_streaming.erb]: 👀

File: gapic-generator/templates/default/service/client/method/def/_request_normal.erb [hunk 6]
```diff
@@ -1,4 +1,15 @@
 <%- assert_locals method -%>
 raise ::ArgumentError, "request must be provided" if request.nil?
+<%- if method.auto_populated_fields.length > 0 -%>
+
+# Auto populate fields for this request.
+<%- end -%>
+<% method.auto_populated_fields.each do |field| %>
+if request.is_a? Hash
+  request[:<%= field %>] = SecureRandom.uuid if request.key?(:<%= field %>) && request[:<%= field %>].empty?
```
[the remark this answers is not in the document]
Comment (aandreassa) [reply gapic-generator/templates/default/service/client/method/def/_request_normal.erb]: Good call, thanks!

File: gapic-generator/templates/default/service/client/method/def/_request_streaming.erb [hunk 7]
```diff
@@ -4,6 +4,17 @@ unless request.is_a? ::Enumerable
   request = request.to_enum
 end
 
+<%- if method.auto_populated_fields.length > 0 -%>
+
+# Auto populate fields for this request.
+<%- end -%>
+<% method.auto_populated_fields.each do |field| %>
+if request.is_a? Hash
+  request[:<%= field %>] = SecureRandom.uuid if request.key? :<%= field %>
```
Comment (dazuma) [inline gapic-generator/templates/default/service/client/method/def/_request_streaming.erb:13]: It needs to be `if !request.key? :<%= field %>` (or probably better, `unless request.key? :<%= field %>`)

Comment (dazuma) [reply gapic-generator/templates/default/service/client/method/def/_request_streaming.erb:13]: Is there a test we could run that would verify this logic?

Comment (aandreassa) [reply gapic-generator/templates/default/service/client/method/def/_request_streaming.erb:13]: Thanks! I am adding one for `echo` specifically and checking some other logic for in service/rest/test/method
/_normal.erb

Comment (aandreassa) [reply gapic-generator/templates/default/service/client/method/def/_request_streaming.erb]: Currently `shared/test/showcase/test_helper.rb` has a comment where we need to update the version for the container image. Locally it passes by building my own binary of `gapic-showcase`. Let's coordinate next week!

**Review by dazuma — APPROVED** [review]

**Pull request merged**